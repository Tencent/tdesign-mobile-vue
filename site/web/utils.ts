export function sortDocs(docs: any[]) {
  let innerDocs = [];
  docs.forEach((item) => {
    let { children } = item;
    if (item.type === 'component') {
      children = item.children.sort((a: any, b: any) => {
        const nameA = a.name.toUpperCase();
        const nameB = b.name.toUpperCase();
        if (nameA < nameB) return -1;
        if (nameA > nameB) return 1;
        return 0;
      });
    }
    if (children) {
      item.children = sortDocs(children);
    }
    innerDocs.push(item);
  });
  return innerDocs;
}

// 过滤小版本号
export function filterVersions(versions = [], deep = 1) {
  const versionMap = Object.create(null);

  versions.forEach((v) => {
    const nums = v.split('.');
    versionMap[nums[deep]] = v;
  });

  return Object.values(versionMap);
}
