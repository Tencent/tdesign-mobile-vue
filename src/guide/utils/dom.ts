import isString from 'lodash/isString';
import isFunction from 'lodash/isFunction';
import { AttachNode } from '../../common';
import { elementInViewport, getWindowScroll, getWindowSize } from '../../shared/dom';
/**
 * 获取元素某个 css 对应的值
 * @param element 元素
 * @param propName css 名
 * @returns string
 */
export function getElmCssPropValue(element: HTMLElement, propName: string): string {
  let propValue = '';

  if (document.defaultView && document.defaultView.getComputedStyle) {
    propValue = document.defaultView.getComputedStyle(element, null).getPropertyValue(propName);
  }

  if (propValue && propValue.toLowerCase) {
    return propValue.toLowerCase();
  }

  return propValue;
}

/**
 * 判断元素是否处在 position fixed 中
 * @param element 元素
 * @returns boolean
 */
export function isFixed(element: HTMLElement): boolean {
  const p = element.parentNode as HTMLElement;

  if (!p || p.nodeName === 'HTML') {
    return false;
  }

  if (getElmCssPropValue(element, 'position') === 'fixed') {
    return true;
  }

  return isFixed(p);
}

/**
 * 获取元素相对于另一个元素的位置（或者说相对于 body）
 * 感谢 `meouw`: http://stackoverflow.com/a/442474/375966
 */
export function getRelativePosition(elm: HTMLElement, relativeElm: HTMLElement = document.body) {
  const { scrollTop, scrollLeft } = getWindowScroll();
  const { top: elmTop, left: elmLeft } = elm.getBoundingClientRect();
  const { top: relElmTop, left: relElmLeft } = relativeElm.getBoundingClientRect();
  const relativeElmPosition = getElmCssPropValue(relativeElm, 'position');

  if (
    (relativeElm.tagName.toLowerCase() !== 'body' && relativeElmPosition === 'relative') ||
    relativeElmPosition === 'sticky'
  ) {
    return {
      top: elmTop - relElmTop,
      left: elmLeft - relElmLeft,
    };
  }

  if (isFixed(elm)) {
    return {
      top: elmTop,
      left: elmLeft,
    };
  }

  return {
    top: elmTop + scrollTop,
    left: elmLeft + scrollLeft,
  };
}

export function getTargetElm(elm: AttachNode): HTMLElement {
  if (elm) {
    let targetElement: HTMLElement = null;
    if (isString(elm)) {
      targetElement = document.querySelector(elm);
    } else if (isFunction(elm)) {
      targetElement = elm() as HTMLElement;
    } else {
      throw new Error('elm should be string or function');
    }
    if (targetElement) {
      return targetElement as HTMLElement;
    }
    if (process?.env?.NODE_ENV !== 'test') {
      throw new Error('There is no element with given.');
    }
  } else {
    return document.body;
  }
}

export function getScrollParent(element: HTMLElement) {
  let style = window.getComputedStyle(element);
  const excludeStaticParent = style.position === 'absolute';
  const overflowRegex = /(auto|scroll)/;

  if (style.position === 'fixed') return document.body;

  for (let parent = element; parent.parentElement; ) {
    parent = parent.parentElement;
    style = window.getComputedStyle(parent);
    if (excludeStaticParent && style.position === 'static') {
      continue;
    }
    if (overflowRegex.test(style.overflow + style.overflowY + style.overflowX)) return parent;
  }

  return document.body;
}

export function scrollToParentVisibleArea(element: HTMLElement) {
  const parent = getScrollParent(element);
  if (parent === document.body) return;
  // !todo 逻辑待验证
  if (elementInViewport(element, parent)) return;
  parent.scrollTop = element.offsetTop - parent.offsetTop;
}

export function scrollToElm(elm: HTMLElement) {
  const rect = elm.getBoundingClientRect();

  if (!elementInViewport(elm)) {
    const winHeight = getWindowSize().height;
    // const top = rect.bottom - (rect.bottom - rect.top);
    window.scrollTo({
      top: rect.top - (winHeight / 2 - rect.height / 2),
      behavior: 'smooth',
    });
  }
}
