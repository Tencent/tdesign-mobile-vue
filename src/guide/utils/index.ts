import {
  getElmCssPropValue,
  isFixed,
  getRelativePosition,
  getTargetElm,
  scrollToParentVisibleArea,
  scrollToElm,
} from './dom';

export { getElmCssPropValue, isFixed, getRelativePosition, getTargetElm, scrollToParentVisibleArea, scrollToElm };
