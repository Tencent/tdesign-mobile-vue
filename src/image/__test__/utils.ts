interface IntersectionObserverCallback {
  (entries: IntersectionObserverEntry[], observer: IntersectionObserver): void;
}

export class MockIntersectionObserver {
  _callback: Function;

  _element!: HTMLElement;

  constructor(callback: Function) {
    this._callback = callback;
  }

  observe(element: HTMLElement) {
    this._element = element;
    this._element.addEventListener('resize', this.trigger);
  }

  unobserve() {
    this._element.removeEventListener('resize', this.trigger);
  }

  disconnect() {
    this._element.removeEventListener('resize', this.trigger);
  }

  trigger = (event: UIEvent) => {
    this._callback([
      {
        isIntersecting: true,
      },
    ]);
  };
}
