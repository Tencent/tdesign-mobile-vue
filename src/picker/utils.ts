import isArray from 'lodash/isArray';
import { PickerColumn } from './type';

export const isMultipleArray = (arr: PickerColumn | PickerColumn[]) => {
  return isArray(arr[0]);
};

export const getPickerColumns = (columns: any): PickerColumn[] => {
  return isMultipleArray(columns) ? columns : [columns];
};
