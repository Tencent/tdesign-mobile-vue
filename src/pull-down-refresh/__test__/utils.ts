import { ComponentPublicInstance, nextTick } from 'vue';
import { VueWrapper, DOMWrapper } from '@vue/test-utils';

function getTouch(el: Element | Window, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5,
  };
}

// 模拟 touch 事件
export function trigger(
  wrapper: VueWrapper<ComponentPublicInstance<any, any, any>> | DOMWrapper<Element> | Element | Window,
  eventName: string,
  x = 0,
  y = 0,
  options: any = {},
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper;
  const touchList = options.touchList || [getTouch(el, x, y)];

  if (options.x || options.y) {
    touchList.push(getTouch(el, options.x, options.y));
  }

  const event = document.createEvent('CustomEvent');
  event.initCustomEvent(eventName, true, true, {});

  Object.assign(event, {
    clientX: x,
    clientY: y,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList,
  });

  el.dispatchEvent(event);

  return nextTick();
}

/**
 * PullDownRefresh 使用了 [`useElementSize()`](https://github.com/vueuse/vueuse/blob/658374fd12/packages/core/useElementSize/index.ts)，
 * 该函数调用了 [`useResizeObserver()`](https://github.com/vueuse/vueuse/blob/658374fd12fbce2ac6127a9fb9bca52fb2137505/packages/core/useResizeObserver/index.ts)。
 *
 * 而 jsdom 中 ResizeObserver 无效，见： https://github.com/jsdom/jsdom/issues/3368
 *
 * @description 模拟实现 ResizeObserver
 * @params callback `new ResizeObserver(callback)` 中的回调，签名如下：
 *
 * ```ts
 * ResizeObserverCallback = (entries: ReadonlyArray<ResizeObserverEntry>, observer: ResizeObserver) => void
 * ```
 */
export class MockResizeObserver {
  _callback: Function;

  _element!: HTMLElement;

  constructor(callback: Function) {
    this._callback = callback;
  }

  observe(element: HTMLElement) {
    this._element = element;
    this._element.addEventListener('resize', this.trigger);
  }

  unobserve() {
    this._element.removeEventListener('resize', this.trigger);
  }

  disconnect() {
    // unmount后监听的元素this._element可能为null，因此先判断元素是否为null，再依据此来是否执行removeEventListener
    if (this._element) {
      this._element.removeEventListener('resize', this.trigger);
    }
    this._element = null;
  }

  /**
   * @description 手动触发 resize，提供参数。本例 `maxBarHeight` 始终为 `useElementSize()` 的默认初始值 `initialSize`（`{ width: 0, height: 0 }`），模拟返回有效值即可
   * @param event 手动触发
   */
  trigger = (event: UIEvent) => {
    const { width = 0, height = 0 } = event as { width?: number; height?: number };
    this._callback([
      {
        contentRect: { width, height },
      },
    ]);
  };
}
