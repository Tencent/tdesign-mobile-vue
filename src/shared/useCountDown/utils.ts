import { TimeData, TdUseCountDownShowTimes } from './type';

/**
 * getRemainTimes
 * @param time
 * @returns
 */
export const getRemainTimes = (time?: number): TimeData => {
  if (!time) {
    return {
      days: 0,
      hours: 0,
      minutes: 0,
      seconds: 0,
      milliseconds: 0,
    };
  }
  const stime = Math.floor(time / 1000); // 秒
  return {
    days: Math.floor(stime / 3600 / 24),
    hours: Math.floor((stime / 3600) % 24),
    minutes: Math.floor((stime / 60) % 60),
    seconds: stime % 60,
    milliseconds: time % 1000,
  };
};

/**
 * fillZero
 * @param num
 * @returns
 */
export const fillZero = (num: number, isMillieconds = false): string | number => {
  if (isMillieconds) {
    if (num >= 100) {
      return num;
    }

    return num >= 10 ? `0${num}` : `00${num}`;
  }

  return num >= 10 ? num : `0${num}`;
};

/**
 * getMark
 * @param format
 * @param type
 * @returns
 */
export const getMark = (format: string, type: string): string => format?.split?.(type)?.[1]?.split?.('')?.[0];

/**
 * getShowTimes
 * @param time
 * @returns
 */
export const getShowTimes = (times: TimeData, format: string, milliseconds = false): TdUseCountDownShowTimes => {
  format = (format || 'DD:HH:mm:ss')?.toUpperCase?.();
  milliseconds && !format.includes(':SSS') && (format = format.concat(':SSS'));
  const showTimes: TdUseCountDownShowTimes = [];
  if (format?.indexOf('DD') > -1) {
    showTimes?.push({
      mark: getMark(format, 'DD'),
      value: fillZero(times?.days),
    });
  }
  if (format?.indexOf('HH') > -1) {
    showTimes?.push({
      mark: getMark(format, 'HH'),
      value: fillZero(times?.hours),
    });
  }
  if (format?.indexOf('MM') > -1) {
    showTimes?.push({
      mark: getMark(format, 'MM'),
      value: fillZero(times?.minutes),
    });
  }
  if (format?.indexOf('SS') > -1) {
    showTimes?.push({
      mark: getMark(format, 'SS'),
      value: fillZero(times?.seconds),
    });
  }
  if (format?.indexOf('SSS') > -1) {
    showTimes?.push({
      mark: getMark(format, 'SSS'),
      value: fillZero(times?.milliseconds, true),
    });
  }

  return showTimes;
};

/**
 * @param {number} targetCount 不小于1的整数，表示经过targetCount帧之后返回结果
 * @return {Promise<number>}
 */
export const getScreenFps = (() => {
  const { requestAnimationFrame, mozRequestAnimationFrame, webkitRequestAnimationFrame } = window as any;
  // 先做一下兼容性处理
  const nextFrame = [requestAnimationFrame, mozRequestAnimationFrame, webkitRequestAnimationFrame]?.find?.((fn) => fn);
  //
  if (!nextFrame) {
    console.error('requestAnimationFrame is not supported!');
    return;
  }
  return (targetCount = 50) => {
    if (targetCount < 1) {
      return;
    }
    let count = 0;
    const beginDate = Date.now();
    return new Promise((resolve) => {
      (function log() {
        nextFrame?.(() => {
          if (++count >= targetCount) {
            const diffDate = Date.now() - beginDate;
            const fps = (count / diffDate) * 1000;
            return resolve(fps);
          }
          log();
        });
      })();
    });
  };
})();
