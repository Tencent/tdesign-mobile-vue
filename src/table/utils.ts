import isFunction from 'lodash/isFunction';
import get from 'lodash/get';
import { CellData, TableColumnClassName, TableRowData } from './type';

export function toString(obj: any): string {
  return Object.prototype.toString.call(obj).slice(8, -1).toLowerCase();
}

export function debounce<T = any>(fn: Function, delay = 200): () => void {
  let timer: ReturnType<typeof setTimeout>;
  return function newFn(this: T, ...args: Array<any>): void {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const context = this;
    clearTimeout(timer);
    timer = setTimeout(() => {
      fn.apply(context, args);
    }, delay);
  };
}

export interface FormatRowAttributesParams {
  row: TableRowData;
  rowIndex: number;
  type: 'body' | 'foot';
}

export function formatClassNames(
  classNames: TableColumnClassName<TableRowData> | TableColumnClassName<TableRowData>[],
  params: CellData<TableRowData>,
) {
  const classes = classNames instanceof Array ? classNames : [classNames];
  const arr: any[] = [];
  for (let i = 0, len = classes.length; i < len; i++) {
    const cls = classes[i];
    if (isFunction(cls)) {
      arr.push(cls(params));
    } else {
      arr.push(cls);
    }
  }
  return arr;
}

export const INNER_PRE_NAME = '@@inner-';

export enum SCROLL_DIRECTION {
  X = 'x',
  Y = 'y',
  UNKNOWN = 'unknown',
}

let preScrollLeft: any;
let preScrollTop: any;

export const getScrollDirection = (scrollLeft: number, scrollTop: number): SCROLL_DIRECTION => {
  let direction = SCROLL_DIRECTION.UNKNOWN;
  if (preScrollTop !== scrollTop) {
    direction = SCROLL_DIRECTION.Y;
  } else if (preScrollLeft !== scrollLeft) {
    direction = SCROLL_DIRECTION.X;
  }
  preScrollTop = scrollTop;
  preScrollLeft = scrollLeft;
  return direction;
};

export function getEditableKeysMap(keys: Array<string | number>, list: any[], rowKey: string) {
  const map: { [key: string | number]: boolean } = {};
  for (let i = 0, len = list.length; i < len; i++) {
    const rowValue = get(list[i], rowKey);
    if (keys.includes(rowValue)) {
      map[rowValue] = true;
    }
  }
  return map;
}

export function getColumnDataByKey(columns: any[], colKey: string): any {
  for (let i = 0, len = columns.length; i < len; i++) {
    if (columns[i].colKey === colKey) return columns[i];
    if (columns[i].children?.length) {
      const t = getColumnDataByKey(columns[i].children, colKey);
      if (t) return t;
    }
  }
  return null;
}

export function getColumnIndexByKey(columns: any[], colKey: string): number {
  for (let i = 0, len = columns.length; i < len; i++) {
    if (columns[i].colKey === colKey) {
      return i;
    }
    if (columns[i].children?.length) {
      const t = getColumnDataByKey(columns[i].children, colKey);
      if (t) return i;
    }
  }
  return -1;
}
