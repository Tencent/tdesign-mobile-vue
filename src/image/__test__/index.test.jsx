import { mount } from '@vue/test-utils';
import { vi, describe, it, expect } from 'vitest';
import { LoadingIcon } from 'tdesign-icons-vue-next';
import { nextTick } from 'vue';
import Image from '../image';
import { MockIntersectionObserver } from './utils';

const prefix = 't';
const name = `${prefix}-image`;
const IMAGE = 'https://tdesign.gtimg.com/site/upload1.png';
const FAIL_IMAGE = 'https://123.jpg';
const fitList = ['fill', 'contain', 'cover', 'none', 'scale-down'];
const shapeList = ['circle', 'round', 'square'];
const positionList = ['top', 'bottom', 'center', 'left', 'right'];

describe('Image', () => {
  beforeAll(() => {
    window.IntersectionObserver = MockIntersectionObserver;
  });

  describe('props', () => {
    it(': lazy', async () => {
      const wrapper = mount(() => <Image src={IMAGE} lazy />);
      await nextTick();
      const $image = wrapper.find(`.${name}__img`);
      expect(wrapper.find(`.${name}__mask`).exists()).toBeFalsy();
      // 触发 IntersectionObserver , 但图片加载完成不会触发 load 回调,
      $image.trigger('resize');
      await nextTick();
      expect($image.attributes('src')).toBe(IMAGE);
      // 手动触发 图片加载完成的回调函数
      await $image.trigger('load');
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.find(`.${name}__mask`).exists()).toBeFalsy();
    });

    it(': alt', async () => {
      const wrapper = mount(() => <Image src={IMAGE} alt="图片" />);
      const $image = wrapper.find(`.${name}__img`);
      expect(wrapper.classes()).toContain(`${name}`);
      expect($image.attributes('alt')).toBe('图片');
    });

    it(': fit', async () => {
      const wrapper = mount(Image, {
        props: {
          fit: '',
          src: IMAGE,
        },
      });
      const $image = wrapper.find(`.${name}__img`);
      // fit = ''
      fitList.forEach((item) => {
        expect($image.attributes('style').includes(`object-fit: ${item}`)).toBeFalsy();
      });

      const fit = 'cover';
      await wrapper.setProps({
        fit,
      });
      // fit = 'square'
      expect($image.attributes('style')).toContain(`object-fit: ${fit}`);
    });

    it(': shape', async () => {
      const wrapper = mount(Image, {
        props: {
          shape: '',
          src: IMAGE,
        },
      });
      const $image = wrapper.findComponent(Image);

      // shape = ''
      shapeList.forEach((item) => {
        expect($image.classes().includes(`${name}--${item}`)).toBeFalsy();
      });

      const shape = 'square';
      await wrapper.setProps({
        shape,
      });
      // shape = 'square'
      expect($image.classes()).toContain(`${name}--${shape}`);
    });

    it(': position', async () => {
      positionList.forEach((position) => {
        const wrapper = mount(() => <Image src={IMAGE} position={position} />);
        const $image = wrapper.find('.t-image__img');
        expect($image.attributes('style')).toContain(`object-position: ${position}`);
      });
    });

    it(': loading', async () => {
      const slots = {
        loading: () => <LoadingIcon />,
      };
      const wrapper = mount(() => <Image src={IMAGE} v-slots={slots} />);
      const $status = wrapper.find(`.${name}__mask`);
      expect($status.exists()).toBeTruthy();
      expect(wrapper.findComponent(LoadingIcon).exists()).toBeTruthy();
    });

    it(': src', async () => {
      const onError = vi.fn();
      const wrapper = mount(() => <Image src="" onError={onError} />);
      await nextTick();
      const $image = wrapper.find(`.${name}__img`);
      // 手动触发 图片加载失败的回调函数
      await $image.trigger('error');
      expect(wrapper.find(`.${name}__mask`).exists()).toBeTruthy();
      // src = ''，不会触发 error
      expect(onError).toBeCalledTimes(0);
    });

    it(': onError', async () => {
      const onError = vi.fn();
      const slots = {
        error: () => '加载失败',
      };
      const wrapper = mount(() => <Image src={FAIL_IMAGE} v-slots={slots} onError={onError} />);
      await nextTick();
      const $image = wrapper.find(`.${name}__img`);
      // 手动触发 图片加载失败的回调函数
      await $image.trigger('error');
      const status = wrapper.find(`.${name}__mask`);
      expect(status.exists()).toBeTruthy();
      expect(status.text()).toBe('加载失败');
      expect(onError).toBeCalledTimes(1);
    });

    it(': onLoad', async () => {
      const onLoad = vi.fn();
      const wrapper = mount(() => <Image src={IMAGE} onLoad={onLoad} />);
      await nextTick();
      const $image = wrapper.find(`.${name}__img`);
      expect($image.attributes('src')).toBe(IMAGE);
      // 手动触发 图片加载完成的回调函数
      await $image.trigger('load');
      expect(onLoad).toBeCalledTimes(1);
    });
  });
});
