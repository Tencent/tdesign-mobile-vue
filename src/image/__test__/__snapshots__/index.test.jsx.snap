// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Image > props > : lazy 1`] = `
<div
  class="t-image t-image--square"
>
  <!---->
  <picture>
    <!---->
    <img
      alt=""
      class="t-image__img"
      src="https://tdesign.gtimg.com/site/upload1.png"
      style="object-fit: fill; object-position: center;"
    />
  </picture>
</div>
`;
